tests/wasm/runtime/events.ts

```diff
@@ -0,0 +1,158 @@
+  ? Uint8Array | string
+  : T extends bigint
+  ? bigint | number
+  : T
+
+export type MetadataProperties<T> = {
+  [K in keyof T]?: T[K] extends (infer S)[]
+    ? MetadataProperties<S>[]
+    : T[K] extends MetadataProperty
+    ? MetadataPropertyLike<T[K]>
+    : MetadataProperties<T[K]>
+}
+
+export type Metadata = MetadataProperties<Omit<Event, 'parameters'>>
+
+export type SimpleValueKind = Exclude<ValueKind, ValueKind.FixedArray | ValueKind.Array | ValueKind.Tuple>
+export type EventValueKind = SimpleValueKind | readonly [SimpleValueKind] | SimpleValueKind[]
+export type Definition = Record<string, EventValueKind>
+
+export type RawValueOf<T> = T extends ValueKind.Address | ValueKind.FixedBytes | ValueKind.Bytes
+  ? Uint8Array | string
+  : T extends ValueKind.Int | ValueKind.Uint
+  ? bigint | number
+  : T extends SimpleValueKind
+  ? Extract<Value, { kind: T }>['data']
+  : never
+export type ValueOf<T> = T extends readonly [infer S]
+  ? RawValueOf<S>[]
+  : T extends SimpleValueKind[]
+  ? unknown[]
+  : RawValueOf<T>
+
+export type Data<T> = {
+  [K in keyof T]: ValueOf<T[K]>
+}
+
+export function toEvent<D extends Definition>(definition: D, data: Data<D>, meta?: Metadata): Event {
+  const parameters: EventParam[] = []
+  const dataProperties = data as Record<string, unknown>
+  for (const [key, kind] of Object.entries(definition)) {
+    parameters.push({ name: key, value: coerceToParameter(dataProperties[key], kind) })
+  }
+
+  return newEvent(parameters, meta)
+}
+
+function isArray(kind: EventValueKind): kind is readonly [SimpleValueKind] {
+  return Array.isArray(kind) && kind.length === 1
+}
+
+function isTuple(kind: EventValueKind): kind is SimpleValueKind[] {
+  return Array.isArray(kind) && kind.length !== 1
+}
+
+function coerceToParameter(data: unknown, kind: EventValueKind): Value {
+  if (isArray(kind)) {
+    return {
+      kind: ValueKind.Array,
+      data: (data as unknown[]).map((data) => coerceToParameter(data, kind[0])),
+    }
+  } else if (isTuple(kind)) {
+    const dataTuple = data as unknown[]
+    if (dataTuple.length !== kind.length) {
+      throw new Error(`tuple ethereum value arity mismatch`)
+    }
+
+    return {
+      kind: ValueKind.Array,
+      data: (data as unknown[]).map((data, i) => coerceToParameter(data, kind[i])),
+    }
+  } else {
+    switch (kind) {
+      case ValueKind.Address:
+        return { kind, data: coerceAddr(data as Address | string) }
+      case ValueKind.FixedBytes:
+      case ValueKind.Bytes:
+        return { kind, data: coerceBytes(data as Uint8Array | string) }
+      case ValueKind.Int:
+      case ValueKind.Uint:
+        return { kind, data: BigInt(data) }
+      default:
+        return { kind, data } as Value
+    }
+  }
+}
+
+function newEvent(parameters: EventParam[], meta?: Metadata): Event {
+  const defaultAddr = `0x${'cd'.repeat(20)}`
+  const defaultHash = `0x${'cd'.repeat(32)}`
```

Comment:
I believe I have seen you use different default Adresses and Hashes before. Any reason we don't just declare a canonical default (e.g. fff..fff) somewhere static?

Reply:
The reason for `0xcd` was another little fun tidbit - MS VC++ compiler generates code to write `0xcdcdcd...` to uninitialized memory in debug builds. It does it to help catch subtle memory bugs such as use after free. I can change it to something more canonical and define it in one place :smile: